File: src/CustomSerialTemperatureSensor.cpp
#include <iostream>
#include <stdio.h>
#include <string.h>
#include <fcntl.h>
#include "CustomSerialTemperatureSensor.h"

CustomSerialTemperatureSensor::CustomSerialTemperatureSensor(ISerialPort &serial_port)
    : serial_port_(serial_port) {}

float CustomSerialTemperatureSensor::readTemperature(const TemperatureUnit unit) {
    std::string response;
    if (unit == TemperatureUnit::Celsius) {
        response = runCommand(COMMAND_GET_TEMPERATURE_CELSIUS);
    } else {
        response = runCommand(COMMAND_GET_TEMPERATURE_FAHRENHEIT);
    }

    // Convert the response string to a float
    return std::stof(response);
}

float CustomSerialTemperatureSensor::readHumidity() {
    std::string response = runCommand(COMMAND_GET_HUMIDITY);
    // Convert the response string to a float with higher precision handling
    try {
        return std::stof(response);
    } catch (const std::exception& e) {
        std::cerr << "Error parsing humidity response: " << e.what() << std::endl;
        return -1.0f; // Return an invalid value to indicate an error
    }
}

std::string CustomSerialTemperatureSensor::readDeviceId() {
    
}

std::string CustomSerialTemperatureSensor::runCommand(const std::string& command) {
    std::string response;
    if (serial_port_.write(command) < 0) {
        std::cerr << "Error writing to serial port: " << strerror(errno) << std::endl;
    }
    else {
        response = serial_port_.readLine();
    }
    return response;
}

File: src/custom_serial_temperature_sensor.py
class CustomSerialTemperatureSensor:
    COMMAND_GET_TEMPERATURE_CELSIUS = "GTC\n"
    COMMAND_GET_TEMPERATURE_FAHRENHEIT = "GTF\n"
    COMMAND_GET_DEVICE_ID = "GID\n"

    def __init__(self, serial_port):
        self.serial_port = serial_port

    def read_temperature(self, unit="Celsius"):
        """Reads temperature data from the serial port."""
        command = (self.COMMAND_GET_TEMPERATURE_CELSIUS
                   if unit == "Celsius" else self.COMMAND_GET_TEMPERATURE_FAHRENHEIT)
        response = self.run_command(command)
        try:
            return float(response)
        except ValueError:
            raise Exception("Invalid temperature data received.")

    def read_device_id(self):
        """TODO: implement this method."""
        return ""

    def run_command(self, command):
        """Sends a command to the serial port and returns the response."""
        self.serial_port.write(command)
        response = self.serial_port.read()
        if not response:
            raise Exception("No response received from the device.")
        return response.strip()
